From 7487eb848c9b15d6672f0f951185f8d13292d2bb Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Wed, 1 Oct 2025 14:42:42 +0800
Subject: [PATCH 237/254] mtk: hostapd: align QCA's prpl Adv-TTLM command
 format

The format is
$ hostapd_cli -i <ifname> advertised_ttlm ieee_link_map=<hex link map>
  map_switch_time=<MST in ms> expected_dur=<ED in ms>
  [link_mapping_size=<size>]

For example:
$ hostapd_cli -i ap-mld-1 advertised_ttlm ieee_link_map=0x5
  map_switch_time=3000 expected_dur=10000 link_mapping_size=0

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 hostapd/ctrl_iface.c  | 25 +++++++++++++++++++++++++
 hostapd/hostapd_cli.c | 14 ++++++++++++++
 2 files changed, 39 insertions(+)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index f08f4ee9a..0c6215994 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -4231,6 +4231,28 @@ static int hostapd_ctrl_iface_set_attlm(struct hostapd_data *hapd, char *cmd,
 				continue;
 		}
 
+		/* Align QCA's command format */
+		if (os_strncmp(token, "ieee_link_map=", 14) == 0) {
+			u16 enabled_links = strtol(token + 14, NULL, 0);
+			if ((enabled_links & valid_links) &&
+			    !(enabled_links & ~valid_links)) {
+				disabled_links = ~enabled_links & valid_links;
+				continue;
+			}
+		}
+
+		if (os_strncmp(token, "map_switch_time=", 16) == 0) {
+			switch_time = atoi(token + 16);
+			if (switch_time > 0 && switch_time <= MAX_SWITCH_TIME_MS)
+				continue;
+		}
+
+		if (os_strncmp(token, "expected_dur=", 13) == 0) {
+			duration = atoi(token + 13);
+			if (duration > 0 && duration <= MAX_DURATION_MS)
+				continue;
+		}
+
 		wpa_printf(MSG_INFO, "CTRL: Invalid SET_ATTLM parameter: %s",
 			   token);
 		return -1;
@@ -7326,6 +7348,9 @@ static int hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 	} else if (os_strncmp(buf, "NEGOTIATED_TTLM ", 16) == 0) {
 		if (hostapd_ctrl_iface_negotiated_ttlm(hapd, buf + 16))
 			reply_len = -1;
+	} else if (os_strncmp(buf, "ADVERTISED_TTLM  ", 16) == 0) {
+		if (hostapd_ctrl_iface_set_attlm(hapd, buf + 16, reply, reply_size))
+			reply_len = -1;
 #endif /* CONFIG_TESTING_OPTIONS */
 #endif /* CONFIG_IEEE80211BE */
 #ifdef CONFIG_SAE
diff --git a/hostapd/hostapd_cli.c b/hostapd/hostapd_cli.c
index 796741fbe..9d8bc823b 100644
--- a/hostapd/hostapd_cli.c
+++ b/hostapd/hostapd_cli.c
@@ -1869,6 +1869,17 @@ static int hostapd_cli_cmd_negotiated_ttlm(struct wpa_ctrl *ctrl, int argc,
        return hostapd_cli_cmd(ctrl, "NEGOTIATED_TTLM", 1, argc, argv);
 }
 
+static int hostapd_cli_cmd_advertised_ttlm(struct wpa_ctrl *ctrl, int argc,
+					   char *argv[])
+{
+       if (argc < 4) {
+               printf("Invalid 'advertised_ttlm command' - needs at least 4 arguments\n");
+               return -1;
+       }
+
+       return hostapd_cli_cmd(ctrl, "ADVERTISED_TTLM", 3, argc, argv);
+}
+
 static int hostapd_cli_cmd_epcs(struct wpa_ctrl *ctrl, int argc, char *argv[])
 {
 	return hostapd_cli_cmd(ctrl, "EPCS", 1, argc, argv);
@@ -2216,6 +2227,9 @@ static const struct hostapd_cli_cmd hostapd_cli_commands[] = {
 		" = Get the Negotiated TTLM Status of the STA" },
 	{ "negotiated_ttlm", hostapd_cli_cmd_negotiated_ttlm, NULL,
 		" = send ttlm test commands" },
+	{ "advertised_ttlm", hostapd_cli_cmd_advertised_ttlm, NULL,
+		"ieee_link_map= map_switch_time= expected_dur= link_mapping_size=\n"
+		"= Trigger advertised TTLM"},
 	{ "epcs", hostapd_cli_cmd_epcs, NULL,
 		" = Control EPCS priority access" },
 	{ "del_mscs", hostapd_cli_cmd_del_mscs, NULL,
-- 
2.45.2

